crates/solver/src/liquidity/zeroex.rs

```diff
@@ -47,17 +50,64 @@ impl ZeroExLiquidity {
             .filter_map(|order| TokenPair::new(order.buy_token, order.sell_token));
         let relevant_pairs = self.base_tokens.relevant_pairs(user_order_pairs);
 
-        let filtered_zeroex_orders = zeroex_orders
+        let order_buckets = self.generate_order_buckets(zeroex_orders, relevant_pairs);
+        let filtered_zeroex_orders = self.get_useful_orders(order_buckets, 5);
+
+        Ok(filtered_zeroex_orders)
+    }
+
+    fn generate_order_buckets(
+        &self,
+        zeroex_orders: impl Iterator<Item = OrderRecord>,
+        relevant_pairs: HashSet<TokenPair>,
+    ) -> OrderBuckets {
+        // divide orders in buckets
+        let mut buckets: OrderBuckets = Default::default();
+        zeroex_orders
             .filter(|record| {
                 match TokenPair::new(record.order.taker_token, record.order.maker_token) {
                     Some(pair) => relevant_pairs.contains(&pair),
                     None => false,
                 }
             })
-            .filter_map(|record| self.record_into_liquidity(record))
-            .collect();
+            .for_each(|order| {
+                let bucket = buckets
+                    .entry((order.order.maker_token, order.order.taker_token))
+                    .or_insert(vec![]);
+                bucket.push(order);
+            });
+        buckets
+    }
 
-        Ok(filtered_zeroex_orders)
+    /// Get the `orders_per_type` best priced and biggest volume orders.
+    fn get_useful_orders(
+        &self,
+        order_buckets: OrderBuckets,
+        orders_per_type: usize,
+    ) -> Vec<Liquidity> {
+        let mut filtered_zeroex_orders = vec![];
+        order_buckets.into_values().for_each(|mut orders| {
+            if orders.len() <= 2 * orders_per_type {
+                filtered_zeroex_orders.extend(orders);
+                return;
+            }
+            // Sorting to have best priced orders at the end of the vector
+            orders.sort_by(|order_1, order_2| {
+                let price_1 = order_1.order.taker_amount as f64 / order_1.order.maker_amount as f64;
+                let price_2 = order_2.order.taker_amount as f64 / order_2.order.maker_amount as f64;
+                price_2.partial_cmp(&price_1).unwrap()
+            });
+            filtered_zeroex_orders.extend(orders.drain(orders.len() - orders_per_type..));
```

Comment:
This part is tricky because there are 3 things that decide the order:
1. which way to divide taker and maker amount
2. which way to partial_cmp (price 1 or price 2 first)
3. which way to take from the sorted vector (start or end)

The way I explained it in my head is that we want those orders that have the maximum maker_amount / taker_amount because we sell the taker amount and receive the maker amount. Based on that I think you can reverse the price calculation and return `price_1.partial_cmp(price_2)`. This feels more natural to me.

Reply:
Changed the logic so that the comparison is between `maker_amount/taker_amount` ratio to have more clarity. Since this is sorted in ascending order and we are picking from the end (drain from the end in the next line), switched to `price_1.partial_cmp(price_2)`.